src/_pytest/monkeypatch.py

```diff
@@ -209,13 +212,23 @@ def delitem(self, dic, name, raising=True):
             self._setitem.append((dic, name, dic.get(name, notset)))
             del dic[name]
 
+    def _warn_if_env_name_is_not_str(self, name):
+        """On Python 2, warn if the given environment variable name is not a native str (#4056)"""
+        if six.PY2 and not isinstance(name, str):
+            warnings.warn(
+                pytest.PytestWarning(
+                    "Environment variable name {!r} should be str".format(name)
+                )
+            )
+
     def setenv(self, name, value, prepend=None):
         """ Set environment variable ``name`` to ``value``.  If ``prepend``
         is a character, read the current environment variable value
         and prepend the ``value`` adjoined with the ``prepend`` character."""
         value = str(value)
```

Comment:
should we do the same for `value`?

Reply:
Not sure, if you guys think it is a good idea I can issue a warning in that case as well.

Reply:
to me it's the same problem as `name` (windows-py2, py3+ crash on non-native-string).  There is potential that someone is depending on the `str(...)` call here unfortunately though :(

Reply:
I was thinking of showing a warning in addition to keep the existing behavior. Python 3 behaves correctly and raises a proper exception already:

```
Python 3.6.3 (v3.6.3:2c5fed8, Oct  3 2017, 18:11:49) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.environ['foo'] = b'1'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Bruno\AppData\Local\Programs\Python\Python36\lib\os.py", line 674, in __setitem__
    value = self.encodevalue(value)
  File "C:\Users\Bruno\AppData\Local\Programs\Python\Python36\lib\os.py", line 730, in check_str
    raise TypeError("str expected, not %s" % type(value).__name__)
TypeError: str expected, not bytes
```

Reply:
so lets warn for bytes on python3 and drop the str when we go python3 only ?

Reply:
actually, the str is a good idea for stuff like setting it to numbers or booleans as well

Reply:
> so lets warn for bytes on python3 and drop the str when we go python3 only ?

Sounds good, will push a new commit shortly.

Reply:
Added a warning when `value` is not `str`, let me know what you guys think. 👍